Source/WebCore/dom/Element.cpp

```diff
@@ -2735,232 +2735,306 @@ void Element::didMoveToNewDocument(Document& oldDocument, Document& newDocument)
     updateEffectiveLangState();
 }
 
-enum class TextDirectionDirective {
-    Invalid,
+// https://html.spec.whatwg.org/multipage/dom.html#attr-dir
+enum class TextDirectionState : uint8_t {
     LTR,
     RTL,
     Auto,
+    Undefined,
 };
 
-static inline TextDirectionDirective parseTextDirection(const AtomString& value)
+static inline TextDirectionState parseTextDirectionState(const AtomString& value)
 {
     if (equalLettersIgnoringASCIICase(value, "ltr"_s))
-        return TextDirectionDirective::LTR;
+        return TextDirectionState::LTR;
     if (equalLettersIgnoringASCIICase(value, "rtl"_s))
-        return TextDirectionDirective::RTL;
+        return TextDirectionState::RTL;
     if (equalLettersIgnoringASCIICase(value, "auto"_s))
-        return TextDirectionDirective::Auto;
-    return TextDirectionDirective::Invalid;
+        return TextDirectionState::Auto;
+    return TextDirectionState::Undefined;
 }
 
-static bool isValidDirValue(const AtomString& value)
+static TextDirectionState elementTextDirectionState(const Element& element)
 {
-    return parseTextDirection(value) != TextDirectionDirective::Invalid;
+    return parseTextDirectionState(element.attributeWithoutSynchronization(dirAttr));
 }
 
-static bool elementAffectsDirectionality(const Element& element)
+static bool elementHasTextDirectionState(const Element& element)
```

Comment:
Maybe we should say `elementHasDefinedTextDirectionState` or `elementHasValidTextDirectionState `?

Reply:
Done.